### DISCUSSION TITLE
At least once VS exactly once

### DISCUSSION THREAD

**Discussion by jarl-dk** [sidekiq/sidekiq#5775] [Q&A] [answered]
On https://github.com/mperham/sidekiq/wiki/Best-Practices#2-make-your-job-idempotent-and-transactional there is a description of a scenario where Sidekiq will perform the ~worker~ job more than once. The scenarios is if the ~Worker~ job fails in which case sidekiq will retry at a later time.

Does Sidekiq guarantee that ~Workers~ jobs that completes successfully runs exactly once. Or does Sidekiq not gurantee this either?

**Comment by mperham** [comment, marked as answer]
FYI the term "worker" is outdated, the modern term is "job".

Even a job which has completed can be re-run. Redis can go down between the point where your job finished but before Sidekiq has acknowledged it in Redis. Sidekiq makes no exactly-once guarantee at all.

**Reply by jarl-dk** [reply]
It would be lovely if this non-guarantee could be emphasized on the page. Because the current example (of failing job) seems to indicate that this is the (only) situation where a job may run more than once.

**Reply by nateberkopec** [reply]
I agree. I've added Mike's reply to that section 👍

**Comment by jarl-dk** [comment]
Follow up question:
When/if sidekiq jobs are performed more than once, will they have the same or different `jid`?
Please also provide this information in the documentation (https://github.com/sidekiq/sidekiq/wiki/Best-Practices#2-make-your-job-idempotent-and-transactional).

**Reply by mperham** [reply]
A different JID is a different job. We are talking about the same job instance, same JID, executing twice.

**Reply by jarl-dk** [reply]
Thanks. Consider emphasize this on the documentation page.
